File: battle_city/examples/nodejs/random.js
class Game {
    constructor(client) {
        this.client = client;
        this.firstTick = false;
        this.start = false;
    }

    receive(data) {
        let color = '\x1B[0m'; // default color
        switch(data.status) {
            case 'data':
                switch(data.action) {
                    case 'move': return; break; // too many data ;_;
                    case 'spawn': color = '\x1B[92m'; break; // green color
                    case 'destroy': color = '\x1B[93m'; break; // orange color
                }
            break;
            case 'ERROR': color = '\x1B[91m'; break; // red color
            case 'OK': color = '\x1B[35m'; break; // purple color
            case 'game':
                color = '\x1B[34m'; // blue color
                switch(data.action) {
                    case 'start': this.start = true; break;
                    case 'over': this.start = false; break;
                }
            break; 
        }

        console.log(color, data, '\x1B[0m');
    }

    loop() {
        if (!this.firstTick) {
            this.send({action: 'greet', name: 'TEST'});
            this.firstTick = true;
        } 

        if (this.start) {
            let action = Math.floor(Math.random() * 3);
            switch(action) {
                case 0:
                    // random speed
                    let speed = Math.round(Math.random() * 2);
                    this.send({action: 'set_speed', speed: speed});
                break;
                case 1:
                    // random direction
                    let index = Math.floor(Math.random() * 4);
                    let direction = ['up', 'down', 'left', 'right'][index];
                    this.send({action: 'rotate', direction: direction});
                break;
                case 2:
                    // SHOT SHOT SHOT
                    this.send({action: 'shot'});
                break;
            }
        }

        this.callSoon(250);
    }

    callSoon(time) {
        setTimeout(this.loop.bind(this), time);
    }

    send(data) {
        let message = JSON.stringify(data);
        this.client.write(message + '\n');
    }

}

let net = require('net');
let client = new net.Socket();
client.setNoDelay();
let game = new Game(client);

client.connect({port: 8888, host: '127.0.0.1'}, function() {
    console.log('\x1B[1mCONNECTED!\x1B[0m');
    game.loop();
});

let buffer = '';
client.on('data', function(raw) {
    let prev, next;
    let message = raw.toString('utf8');
    while ((next = message.indexOf('\n', prev)) > -1) {
      buffer += message.substring(prev, next);

      let data = JSON.parse(buffer);
      game.receive(data);

      buffer = '';
      prev = next + 1;
    }
    buffer += message.substring(prev);
});

client.on('close', function() {
	console.log('Connection closed');
});

File: battle_city/examples/ruby/random.rb
require 'socket'
require 'json'
require 'time'


class Game
    def initialize(socket)
        @start = false
        @first_tick = false
        @socket = socket
        @timestamp = 0.0
        @prng = Random.new
    end

    def receive(data)
        # ruby with no external libs doesnt have async code
        # so I must check timestamp and run loop
        
        new_timestamp = Time.now.to_f
        if new_timestamp - @timestamp > 0.25
            tick
            @timestamp = new_timestamp
        end

        color = "\e[0m"  # default color
        case data['status'] 
        when 'data'
            case data['action']
            when 'move'
                return
            when 'spawn'
                color = "\e[92m"  # green color
            when 'destroy'
                color = "\e[93m"  # orange color
            end
        when 'ERROR'
            color = "\e[91m"  # red color
        when 'OK'
            color = "\e[35m"  # purple color
        when 'game'
            color = "\e[34m"  # blue color
            case data['action']
            when 'start'
                @start = true
            when 'over'
                @start = false
            end
        end
        print color, data
        puts "\e[0m"
    end

    def tick
        if !@first_tick
            send 'action' => 'greet', 'name' => 'RUBY'
            @first_tick = true
        end

        if @start 
            action = @prng.rand(0..2)
            case action
            when 0  # random speed
                speed = @prng.rand(0..2)
                send 'action' => 'set_speed', 'speed' => speed
            when 1  # random direction
                index = @prng.rand(0..3)
                direction = ['up', 'down', 'left', 'right'][index]
                send 'action' => 'rotate', 'direction' => direction
            when 2  # SHOT SHOT
                send 'action' => 'shoot'
            end
        end
    end

    def send(data)
        @socket.puts(data.to_json)
    end
end

socket = TCPSocket.open('127.0.0.1', 8888)
game = Game.new(socket)
game.tick

while line = socket.gets
    game.receive(JSON.parse!(line))
end
socket.close
